json_rpc/clients/httpclient.nim

```diff
@@ -65,17 +65,38 @@ method call*(client: RpcHttpClient, name: string,
   let
     id = client.getNextId()
     reqBody = $rpcCallNode(name, params, id)
-    req = HttpClientRequestRef.post(client.httpSession,
-                                    client.httpAddress.get,
-                                    body = reqBody.toOpenArrayByte(0, reqBody.len - 1),
-                                    headers = headers)
-    res =
-      try:
-        await req.send()
-      except CancelledError as e:
-        raise e
-      except CatchableError as e:
-        raise (ref RpcPostError)(msg: "Failed to send POST Request with JSON-RPC.", parent: e)
+
+  var req: HttpClientRequestRef
+  var res: HttpClientResponseRef
+
+  defer:
+    # BEWARE!
+    # Using multiple defer statements in this function or multiple
+    # try/except blocks within a single defer statement doesn't
+    # produce the desired run-time code, so we use slightly bizzare
+    # code to ensure the exceptions safety of this function:
+    try:
+      var closeFutures = newSeq[Future[void]]()
+      if req != nil: closeFutures.add req.closeWait()
+      if res != nil: closeFutures.add res.closeWait()
+      if closeFutures.len > 0: await allFutures(closeFutures)
+    except CatchableError as err:
```

Comment:
exceptions in finally handlers are problematic in general - does `closeWait` really raise? if so, I suspect that needs fixing so that it never raises, or it becomes quite difficult to reason about shutdown in general